Enable kubectl to edit all exported kubeconfig settings with json tags

mpuckett159 opened the pull request:
/kind feature
/hold

#### What this PR does / why we need it:
This PR enables users to set and unset all configuration keys that correspond to config settings with json tags

#### Which issue(s) this PR fixes:
fixes kubernetes/kubectl#1093

#### Does this PR introduce a user-facing change?
```release-note
users can now edit all settings with json tags in the kubeconfig
```

### Examples
#### Set auth-provider name
```bash
❯ ./_output/local/bin/darwin/amd64/kubectl config set users.foo.auth-provider.name test
Property "users.foo.auth-provider.name" set.
```
Resulting config:
```
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users:
- name: foo
  user:
    auth-provider:
      config: null
      name: test
```
#### Unset auth-provider name
```bash
❯ ./_output/local/bin/darwin/amd64/kubectl config unset users.foo.auth-provider.name     
error: cannot unset the name of auth-provider, must update or unset auth-provider itself
```
Original config:
```
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users:
- name: foo
  user:
    auth-provider:
      config: null
      name: test
```
Resulting config:
```
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users:
- name: foo
  user:
    auth-provider:
      config: null
      name: ""
```
#### Set auth-provider config refresh-token
```bash
❯ ./_output/local/bin/darwin/amd64/kubectl config set users.foo.auth-provider.config.refresh-token test
Property "users.foo.auth-provider.config.refresh-token" set.
```
Resulting config:
```
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users:
- name: foo
  user:
    auth-provider:
      config:
        refresh-token: test
      name: ""
```
#### Unset auth-provider config refresh-token
```bash
❯ ./_output/local/bin/darwin/amd64/kubectl config unset users.foo.auth-provider.config.refresh-token
Property "users.foo.auth-provider.config.refresh-token" unset.
```
Original config:
```
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users:
- name: foo
  user:
    auth-provider:
      config:
        refresh-token: test
      name: ""
```
Resulting config:
```
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users:
- name: foo
  user:
    auth-provider:
      config: {}
      name: ""
```
#### Unset auth-provider config
```bash
❯ ./_output/local/bin/darwin/amd64/kubectl config unset users.foo.auth-provider.config
Property "users.foo.auth-provider.config" unset.
```
Original config:
```
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users:
- name: foo
  user:
    auth-provider:
      config:
        refresh-token: test
      name: ""
```
Resulting config:
```
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users:
- name: foo
  user:
    auth-provider:
      config: null
      name: ""
```
#### Set new auth-provider name
```bash
❯ ./_output/local/bin/darwin/amd64/kubectl config set users.test.auth-provider.name test
Property "users.test.auth-provider.name" set.
```
Original config:
```
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users:
- name: foo
  user:
    auth-provider:
      config: null
      name: oidc
```
Resulting config:
```
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users:
- name: foo
  user:
    auth-provider:
      config: null
      name: test
```
#### Set new auth-provider with config refresh-token
```bash
❯ ./_output/local/bin/darwin/amd64/kubectl config set users.test.auth-provider.config.refresh-token test
Property "users.test.auth-provider.config.refresh-token" set.
```
Original config:
```
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users:
- name: foo
  user:
    auth-provider:
      config:
        refresh-token: test
      name: ""
```
Resulting config:
```
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users:
- name: foo
  user:
    auth-provider:
      config:
        refresh-token: testnew
      name: ""
```
#### Set auth-provider with no key
```bash
❯ ./_output/local/bin/darwin/amd64/kubectl config set users.test.auth-provider test
error: unable to locate path auth-provider
```
#### Unset auth-provider
```bash
❯ ./_output/local/bin/darwin/amd64/kubectl config unset users.foo.auth-provider
Property "users.foo.auth-provider" unset.
```
Original config:
```
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users:
- name: foo
  user:
    auth-provider:
      config:
        refresh-token: testnew
      name: ""
```
Resulting config:
```
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users:
- name: foo
  user: {}
```
#### Set new exec args
```bash
❯ ./_output/local/bin/darwin/amd64/kubectl config set users.foo.exec.args test1,test2,test3
Property "users.foo.exec.args" set.
```
Resulting config:
```
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users:
- name: foo
  user:
    exec:
      args:
      - test1
      - test2
      - test3
      command: ""
      env: null
      provideClusterInfo: false
```
#### Add new exec arg
```bash
❯ ./_output/local/bin/darwin/amd64/kubectl config set users.foo.exec.args test4+           
Property "users.foo.exec.args" set.
```
Original config:
```
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users:
- name: foo
  user:
    exec:
      args:
      - test1
      - test2
      - test3
      command: ""
      env: null
      provideClusterInfo: false
```
Resulting config:
```
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users:
- name: foo
  user:
    exec:
      args:
      - test1
      - test2
      - test3
      - test4
      command: ""
      env: null
      provideClusterInfo: false
```
#### Remove exec args
```bash
❯ ./_output/local/bin/darwin/amd64/kubectl config set users.foo.exec.args test2-
Property "users.foo.exec.args" set.
```
Original config:
```
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users:
- name: foo
  user:
    exec:
      args:
      - test1
      - test2
      - test3
      - test4
      command: ""
      env: null
      provideClusterInfo: false
```
Resulting config:
```
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users:
- name: foo
  user:
    exec:
      args:
      - test1
      - test3
      - test4
      command: ""
      env: null
      provideClusterInfo: false
```
#### Overwrite existing exec args
```bash
❯ ./_output/local/bin/darwin/amd64/kubectl config set users.foo.exec.args test5,test6
Property "users.foo.exec.args" set.
```
Original config:
```
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users:
- name: foo
  user:
    exec:
      args:
      - test1
      - test3
      - test4
      command: ""
      env: null
      provideClusterInfo: false
```
Resulting config:
```
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users:
- name: foo
  user:
    exec:
      args:
      - test5
      - test6
      command: ""
      env: null
      provideClusterInfo: false
```
#### Set new exec environment variable
```bash
❯ ./_output/local/bin/darwin/amd64/kubectl config set users.foo.exec.env.name.test value:val1
Property "users.foo.exec.env.name.test" set.
```
Resulting config:
```
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users:
- name: foo
  user:
    exec:
      args: null
      command: ""
      env:
      - name: test
        value: val1
      provideClusterInfo: false
```
#### Update existing exec environment variable
```bash
❯ ./_output/local/bin/darwin/amd64/kubectl config set users.foo.exec.env.name.test value:val2
Property "users.foo.exec.env.name.test" set.
```
Original config:
```
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users:
- name: foo
  user:
    exec:
      args: null
      command: ""
      env:
      - name: test
        value: val1
      provideClusterInfo: false
```
Resulting config:
```
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users:
- name: foo
  user:
    exec:
      args: null
      command: ""
      env:
      - name: test
        value: val2
      provideClusterInfo: false
```
#### Set new act-as-user-extra
```bash
❯ ./_output/local/bin/darwin/amd64/kubectl config set users.foo.act-as-user-extra.test1 val1,val2,val3
Property "users.foo.act-as-user-extra.test1" set.
```
Resulting config:
```apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users:
- name: foo
  user:
    as-user-extra:
      test1:
      - val1
      - val2
      - val3
```
#### Add new act-as-user-extra
```bash
❯ ./_output/local/bin/darwin/amd64/kubectl config set users.foo.act-as-user-extra.test1 val4+
Property "users.foo.act-as-user-extra.test1" set.
```
Original config:
```
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users:
- name: foo
  user:
    as-user-extra:
      test1:
      - val1
      - val2
      - val3
```
Resulting config:
```
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users:
- name: foo
  user:
    as-user-extra:
      test1:
      - val1
      - val2
      - val3
      - val4
```
#### Remove act-as-user-extra
```bash
❯ ./_output/local/bin/darwin/amd64/kubectl config set users.foo.act-as-user-extra.test1 val3-
Property "users.foo.act-as-user-extra.test1" set.
```
Original config:
```
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users:
- name: foo
  user:
    as-user-extra:
      test1:
      - val1
      - val2
      - val3
      - val4
```
Resulting config:
```
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users:
- name: foo
  user:
    as-user-extra:
      test1:
      - val1
      - val2
      - val4
```
#### Overwrite existing act-as-user-extra
```bash
❯ ./_output/local/bin/darwin/amd64/kubectl config set users.foo.act-as-user-extra.test1 val8,val9,val0 
Property "users.foo.act-as-user-extra.test1" set.
```
Original config:
```
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users:
- name: foo
  user:
    as-user-extra:
      test1:
      - val1
      - val2
      - val4
```
Resulting config:
```
apiVersion: v1
clusters: null
contexts: null
current-context: ""
kind: Config
preferences: {}
users:
- name: foo
  user:
    as-user-extra:
      test1:
      - val0
      - val8
      - val9
```

mpuckett159 committed:
Enable editing auth-provider information

mpuckett159 committed:
Add test for unsetting auth-provider info

liggitt commented:
/cc @deads2k

mpuckett159 commented:
I'm realizing this is even more of a work in progress than I originally thought. You can set and unset keys under an existing auth-provider config, but you can't set a new auth-provider config entirely. I made this originally because I wanted to solicit some help because I'm having a hard time groking the reflect package, but I wanted to at least have part of it working properly before I started badgering people.

mpuckett159 committed:
Ensure path length is checked properly

Previously the path length wasn't checked to make sure that it wasn't too long. All steps in the path beyond the auth-provider config key were ignored. Now an appropriate error is returned if this is the case.

mpuckett159 committed:
Allow setting new auth-provider with name key

So far we now are able to:
 * set new config key in existing auth-provider config
 * unset new config key in existing auth-provider config
 * set new auth-provider config with name key

Can not:
 * set new auth-provider config with config key

mpuckett159 committed:
Enable setting new auth-provider config keys with new user

mpuckett159 committed:
Case handling for improper set/unset path arguments

mpuckett159 committed:
go fmt commit

mpuckett159 commented:
Ok this feels more complete now. Updated the original PR message with examples of what does and doesn't work and how the file is updated by the commands.

mpuckett159 committed:
Clarify comments and logging statements

mpuckett159 commented:
/unhold

mpuckett159 commented:
I have realized that this behavior is actually covered by the `set-credentials` subcommand. Not sure if I should update this to only be an update to the logging messages to be more explicit, or if allowing users to set using the `set` subcommand with the paths laid out above is considered useful enough.

eddiezane commented:
/assign

eddiezane commented:
/ok-to-test

mpuckett159 committed:
Fix test failures

Due to updates to to the logging a previously existing test (TestParseWithNoMatchingValue) needs to be updated for the expected response.

Fix test I made that I forgot to update before pushing to the PR.

mpuckett159 commented:
/retest

mpuckett159 committed:
go fmt

mpuckett159 committed:
Output more details on DeepEqual test fail

mpuckett159 committed:
AuthInfo.LocationOfOrigin work around

Because the edited config for AuthInfo sets the pwd for the test as the LocationOfOrigin, we must work around this by setting the values to be equal to each other, taking care to only set the keys necessary so we can test various scenarios.

mpuckett159 committed:
Add tests for AuthProviderConfig settings

mpuckett159 commented:
/retest-required

mpuckett159 commented:
Hey @eddiezane (and/or anyone else that sees this) just wanted to see if there was anything else needed from me to get things moving. I have a subsequent PR I'd like to make to get the rest of the possible keys in the kubeconfig editable by `kubectl config set` but it has some dependency on this PR so just wanted to get this one closed out if it's acceptable to everyone.

mpuckett159 committed:
Remove type dependencies

mpuckett159 committed:
more exhaustive tests

mpuckett159 committed:
Enable ExecConfig settings

mpuckett159 committed:
Enable editing more slice types

mpuckett159 committed:
Support setting certain map types

mpuckett159 committed:
Fix --set-raw-bytes test

mpuckett159 committed:
go fmt

mpuckett159 committed:
Complete removal of directly defining types in set.go

mpuckett159 committed:
Update tests to work with new navigation and value formatting

mpuckett159 commented:
/test pull-kubernetes-e2e-kind

mpuckett159 commented:
Alright updated the PR message to fully flesh out all new keys made available to set via `kubectl config set`.

Some notes:
 * I'm using a `val1,val2+` and `val1,val2-` notation to allow users add new values to existing lists and subtract them from existing lists.
 * Exec environment variables requires using slightly unorthodox arguments:
   - `users.foo.exec.env.name.test` specifies a field (`name`) to search for a matching value (`test`) in the slice of structs
   - `value:val1` specifies the field (`value`) to set to the input value (`val1`)

Other than that it's pretty much exactly how the existing `config set` notation works.

Also big thanks to @eddiezane for helping me get this into a better state.

mpuckett159 committed:
throw error on empty map keys

Identified a sharp edge, if no map key is provided it will automatically be set to an empty string instead of erroring. This is undesirable because it will likely be the result of user error rather than what a user truly wants to do. Added a check to ensure this is not the case and return an error.

eddiezane commented on staging/src/k8s.io/kubectl/pkg/cmd/config/set_test.go:
```diff
@@ -24,6 +24,8 @@ import (
 
 	"reflect"
 
+	"github.com/google/go-cmp/cmp"
```
nit: 

```
import (
builtin

3rdparty

k8s
)
```

eddiezane commented on staging/src/k8s.io/kubectl/pkg/cmd/config/set_test.go:
```diff
@@ -54,6 +57,1183 @@ func TestSetConfigCurrentContext(t *testing.T) {
 	test.run(t)
 }
 
+func TestSetConfigAuthProviderName(t *testing.T) {
```
Can we please make these table driven tests to cut down on duplication?

mpuckett159 replied:
Hopefully what I did is what you wanted, but it didn't really seem to save us many lines. Apologies if I've misunderstood.

eddiezane commented on staging/src/k8s.io/kubectl/pkg/cmd/config/set.go:
```diff
@@ -254,9 +334,100 @@ func modifyConfig(curr reflect.Value, steps *navigationSteps, propertyValue stri
+}
 
-	panic(fmt.Errorf("unrecognized type: %v", actualCurrValue))
+func editStringSlice(slice []string, input string) []string {
```
This will need to make it into the command help text.

eddiezane commented on staging/src/k8s.io/kubectl/pkg/cmd/config/set.go:
```diff
@@ -151,20 +152,40 @@ func modifyConfig(curr reflect.Value, steps *navigationSteps, propertyValue stri
+					currMapValue = reflect.New(mapValueType)
+					actualCurrValue.SetMapIndex(mapKey, reflect.Indirect(currMapValue))
+				}
+				currentSliceValue := actualCurrValue.MapIndex(mapKey).Interface().([]string)
```
Ah I see your comment about this field now. Still not sure if we can assert this though?

eddiezane reviewed (changes requested):
This is looking much closer!

mpuckett159 committed:
Remove AuthProviderConfig and ExecConfig cases in navigation parser

mpuckett159 commented on staging/src/k8s.io/kubectl/pkg/cmd/config/navigation_step_parser.go:
```diff
@@ -76,6 +83,40 @@ func newNavigationSteps(path string) (*navigationSteps, error) {
+
+		case reflect.Ptr:
+			// Similar to the modifyConfig function, because we can't easily walk the AuthProviderConfig or ExecConfig types we must manually work with them
+			switch currType {
```
Whoops, didn't even notice after all the changes that the cases were identical. Will collapse this, thank you.

mpuckett159 committed:
Provide error to user on malformed ExecEnv value

mpuckett159 committed:
Add type check to map of slices case

This is to protect against invalid type assertions

mpuckett159 committed:
Return error instead of panicking

mpuckett159 committed:
Reorder imports appropriately

mpuckett159 committed:
go fmt

mpuckett159 committed:
Add helper function to get yaml (json) tags from structs

mpuckett159 committed:
Add helper func to find struct field index by a given name

mpuckett159 committed:
Simplify getStructByFieldName args, update comment

mpuckett159 commented on staging/src/k8s.io/kubectl/pkg/cmd/config/set.go:
```diff
@@ -254,9 +334,100 @@ func modifyConfig(curr reflect.Value, steps *navigationSteps, propertyValue stri
+	return -1
+}
 
+func dedupeStringSlice(slice []string) []string {
```
Ah this is so nice. Thank you for pointing me to this.

mpuckett159 committed:
Utilize k8s.io/apimachinery/pkg/util/sets for string manipulation

mpuckett159 committed:
Updated help text to account for new functionality

mpuckett159 committed:
Move to table based testing

mpuckett159 committed:
go fmt

eddiezane commented:
Cross posting test cases from Slack.

```

mpuckett159 committed:
Remove list ordering as it can mess up intended user configs

An example here would be if the user provides an ordered set of arguments for `users.foo.exec.args` and expects them to remain ordered to work properly during the exec process.

Additionally updated the example text to note an edge case where the first arg in the list starts with a - or --.

Remove k8s.io/apimachinery/pkg/util/sets dependency as well as it was only being used for the list ordering function.

mpuckett159 committed:
Add deduplication flag and test cases

mpuckett159 committed:
Remove unnecessary checks in navigation step parsing

They are unnecessary and more importantly cause issues with ensuring that users are actually providing valid entires.

mpuckett159 committed:
Test for setting a bad key giving an error

mpuckett159 committed:
Fix issue with adding new exec env vars

mpuckett159 committed:
Properly check for bad paths in slice of strings case

mpuckett159 committed:
Add test to validate adding new exec env var to existing config

mpuckett159 committed:
Add ability to delete exec env vars and tests

mpuckett159 commented:
At the point of this comment bad paths should return errors properly again.

Fixed issues with exec env vars not getting set properly, with full ability to set new, update existing, add in addition to existing, and delete existing exec env vars.

mpuckett159 commented:
/retest

mpuckett159 committed:
Change deletion syntax of exec env vars

This more closely follows the traditional shell env var syntax and process

mpuckett159 committed:
Fix up tests with more consistently formatted messages

mpuckett159 commented:
Re worked the exec env var syntax so that when a user wants to set one to an empty string it insets the env var, to more closely align with the process of unsetting env vars in shell environments, e.g.:

`kubectl config set users.foo.exec.env.name.test value:`

Will result in the exec env var with the name test will be deleted from the list of it exists.

mpuckett159 commented:
/retest

mpuckett159 committed:
Resolve testing errors

mpuckett159 committed:
improve error handling

mpuckett159 committed:
fix navigation step parser issues

Add more cases and checks to ensure that steps are valid.

Update tests to take new error messages into account.

mpuckett159 committed:
Improve error handling and messaging

mpuckett159 committed:
fix unit test issue

mpuckett159 commented:
/retest

mpuckett159 commented:
/retest

eddiezane commented:
/assign @KnVerey @soltysh

soltysh reviewed (changes requested):
/hold 
I had a closer look at this PR and my biggest worry I have here is that we're expanding DSL which is strictly limited to config modifications only. In parallel we already have tools allowing us to express similar (maybe not identical) operations through JSONPath see https://kubernetes.io/docs/reference/kubectl/jsonpath/ which kubectl users are already familiar with because it's heavily used when you invoke `kubectl get -o jsonpath=...`. 

I'm proposing that we start the deprecation of the old mechanism in favor of existing JSONPath support. This would work such that the user's input would firstly be matched against JSONPath, then through the old mechanism (without expansion) and only then fail. in the long run I think this will benefit the product and make it more consistent across the board. 

I'm happy to talk about it more in the next SIG-CLI call.

mpuckett159 commented:
Thanks for taking a look @soltysh , the point about not having to support a custom language makes sense and I agree, I'll work on a rewrite to suppose using JSONPath. Do we think it's worth keeping what has been written as a temporary extension of the existing dot delimited pathing, but just be aware it'll get deprecated eventually anyway, or do we want to axe the whole thing and restart using JSONPath only? My feelings won't be hurt either way.

mpuckett159 commented:
Alright after playing with this (jsonpath) a little bit, this may not be ideal if we're looking to avoid implementing a domain specific language. The `clientcmdapi.Config` object is not really a 1:1 representation of the yaml/json that is written out to the file. An example:

```
users:
- name: experimenter
  user:
    password: some-password
    username: exp
```

To access the username field here, it would be expected by someone familiar with jsonpath that the notation would be `{.users[?(@.name=="experimenter")].user.username`, but because of how the `clientcmdapi.Config` object is structured, the reality is that the proper notation will be `{.users.experimenter.username}`. Now, this is obviously a lot easier to read and write (in fact, we can most likely just rip and replace our existing navigation step parsing and existing logic with this for the most part by simply wrapping the provided property name with `{.<arg>}` and achieve the exact same results) but this still basically makes this a DSL.

I don't really think this is much of a problem personally, but thought I would detail it so that everyone was aware.

If this isn't what we're looking for, we would either need to just unpack the raw json/yaml as provided in the config file to allow users to supply normally formatted jsonpath strings to query and set fields, which would force us to rewrite a lot logic to make this work, especially if we expect this to support setting values along with all their required parent attributes, the way it works now.

mpuckett159 commented:
Copying over from slack for posterity.

I’ve been working on implementing jsonpath for the config set subcommand, and because of the way the Config struct in [k8s.io/client-go/tools/clientcmd/api](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/clientcmd/api/types.go) is implemented, it’s not a one-to-one representation of the actual json/yaml of the way the config file is written.

Example:
```
users:
- name: experimenter
  user:
    password: some-password
    username: exp
```
So to get at the username value here, if we use the Config struct, the jsonpath would be `{.users.experimenter.username} `instead of the expected (defined [here](https://kubernetes.io/docs/reference/kubectl/jsonpath/)) `{.users[?(@.name=="experimenter")].user.username}`.

So three options that I see:
- umarshal the config json/yaml onto an interface and go from there, losing all our typing but following the normal jsonpath syntax
- make a parallel config struct that is one-to-one with the json/yaml to unmarshal onto
- still implement a domain specific language but just wrap it in jsonpath

mpuckett159 commented:
Reminder for myself for later:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/clientcmd/api/v1/types.go

mpuckett159 commented:
Moving work for enabling jsonpath support to a new branch so I can preserve this one for a portfolio or whatever.

https://github.com/kubernetes/kubernetes/pull/110392
